In [416]:
from utils import setup_env_path

setup_env_path()

%load_ext autoreload
%autoreload 2

Former working directory:  /Users/ambroisebertin/Desktop/prog/prog_abeilles/fil-rouge-pollinisateurs
Current working directory:  /Users/ambroisebertin/Desktop/prog/prog_abeilles/fil-rouge-pollinisateurs
The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [417]:
import pandas as pd
import numpy as np
from tqdm import tqdm
import missingno as msno

In [ ]:
API_keys = [
    "2b10Me1HF0rfjoGWCseolNa8e",
    "2b10Xs7brWPuBdRTSeWx9V7HJu",
    "2b10umm8L2jTYjlWrJKPfshJ0O",
    "2b10sEg5pSsrtVT372XqlrzcLe",
    "2b10LDWWlXJwuS6FleaqfK6Rke",
    "2b10BoBZKV8WqZx9XCYulKvdu",
    "2b10U9ilY13D9JvF70m7WjK1kO",
    "2b101ARTX7UKqG1hqaeL9qav",
    "2b10Aun0VZ0L3WxrRmMRtx6rsO",
    "2b10PHd7xZZfWqzPghtBrYY6Me",
    "2b10uRbRzoHPIDpZaY3ID9o4e",
    "2b104GXJh341DO2S60qo3ABkI",
]

On a 500 requêtes PlantNet par clef API chaque jour.

## Préparation des données

### sauter cette partie si les fichiers csv existent déjà dans temporary_data/plantes

A l'origine cette partie a été utilisée pour générer les datasets qui sont maintenant présents dans data/temporary_data/plantes. Ainsi, à moins de devoir recommencer de zéro, il n'est plus nécessaire d'executer ce code (mais il peut aider à comprendre le contenu des deux datasets construits).

In [473]:
# Import spipoll.csv as a pandas dataframe
spipoll = pd.read_csv("././data/spipoll.csv",low_memory=False)

# Extract relevant columns from the spipoll dataframe
plantes = spipoll[['collection_id', 'plante_sc', 'plante_fr',
       'plante_precision', 'plante_inconnue', 'plante_caractere',
       'photo_fleur', 'photo_plante', 'photo_feuille']]

# Shrinking the data by grouping by collection_id
# Keep only the first row for each unique value of collection_id
plantes = plantes.drop_duplicates(subset='collection_id', keep='first')

plantes.shape

(75372, 9)

In [474]:
values_to_check = ["Je ne sais pas", "Plante inconnue"]

In [475]:
plantes_sc_unlabelled = plantes.loc[
    (
        (plantes["plante_sc"].isna() | plantes["plante_sc"].isin(values_to_check))
    )
    | ((plantes["plante_inconnue"] == 1.0))
]

# creation du dataframe contenant les valeurs de plantes mais pas de plantes_sc_unlabelled
plantes_sc_labelled = plantes[~plantes.index.isin(plantes_sc_unlabelled.index)]

plantes_sc_labelled.shape

(63465, 9)

In [476]:
# verification des dimensions des deux df par rapport à la dim de plantes
plantes_sc_labelled.shape[0] + plantes_sc_unlabelled.shape[0] == plantes.shape[0]

True

In [477]:
plantes_precision_unlabelled = plantes_sc_unlabelled.loc[
    (
        (plantes["plante_precision"].isna() | plantes["plante_precision"].isin(values_to_check))
    )
]

# creation du dataframe contenant les valeurs de plantes_sc_unlabelled mais pas de plantes_precision_unlabelled
plantes_precision_labelled = plantes_sc_unlabelled[~plantes_sc_unlabelled.index.isin(plantes_precision_unlabelled.index)]

plantes_precision_labelled.shape

(9619, 9)

In [478]:
# verification des dimensions des deux df par rapport à la dim de plantes_sc_unlabelled
plantes_precision_labelled.shape[0] + plantes_precision_unlabelled.shape[0] == plantes_sc_unlabelled.shape[0]

True

Compression de l'information étiquettée redondante :

In [ ]:
# group plantes_sc_labelled by unique values of plante_sc
plantes_sc_labelled = plantes_sc_labelled.drop_duplicates(subset='plante_sc', keep='first')

# display the dimension of the dataframe
plantes_sc_labelled.shape

In [ ]:
# group plantes_precision_labelled by unique values of plante_precision
plantes_precision_labelled = plantes_precision_labelled.drop_duplicates(subset='plante_precision', keep='first')

# display the dimension of the dataframe
plantes_precision_labelled.shape

In [ ]:
# ajoute trois colonnes après plante_sc dans le dataframe plantes_sc_labelled "Famille", "Genre" et "Espèce" entre la colonne plante_sc et plante_fr
plantes_sc_labelled.insert(2, "Famille", "")
plantes_sc_labelled.insert(3, "Genre", "")
plantes_sc_labelled.insert(4, "Espece", "")

# ajoute trois colonnes après plante_sc dans le dataframe plantes_precision_labelled "Famille", "Genre" et "Espèce" entre la colonne plante_sc et plante_fr
plantes_precision_labelled.insert(2, "Famille", "")
plantes_precision_labelled.insert(3, "Genre", "")
plantes_precision_labelled.insert(4, "Espece", "")

# ajoute trois colonnes après plante_sc dans le dataframe plantes_precision_unlabelled "Famille", "Genre" et "Espèce" entre la colonne plante_sc et plante_fr
plantes_precision_unlabelled.insert(2, "Famille", "")
plantes_precision_unlabelled.insert(3, "Genre", "")
plantes_precision_unlabelled.insert(4, "Espece", "")

## Appel à l'API

In [ ]:
from data_quality.plant_treatment.plantnet_api import PlantNetPredictor

def call_API(df, iloc, api_key_index=0):
    
    predictor = PlantNetPredictor(key = API_keys[api_key_index])
    prediction = None

    try:
        prediction = predictor.predict(imageURL= df['photo_fleur'].iloc[iloc],
                        organs = "auto",
                        includeRelatedImages=False)
    except Exception as e:
        if api_key_index + 1 < len(API_keys):
            return call_API(df, iloc, api_key_index + 1)
        else:
            print(f"Error: {e}")
            return ["erreur_API", "erreur_API", "erreur_API"]

    try:
        famille = prediction["results"][0]["species"]["family"]["scientificNameWithoutAuthor"] # famille
        genre = prediction["results"][0]["species"]["genus"]["scientificNameWithoutAuthor"] # genre
        espece = prediction["results"][0]["species"]["scientificNameWithoutAuthor"] # espece
    
    except IndexError:
        print("IndexError: list index out of range")
        return ["erreur_API", "erreur_API", "erreur_API"]

    return [famille, genre, espece]

In [ ]:
def remplir_tableau(df, num):
    
    for i in tqdm(range(num)):
        start_index = 0

        for j, val in enumerate(df["Famille"]):
            if val == "":
                start_index = j
                break
        else:
            print("Le dataframe est déjà rempli.")
            return df

        try:
            df.iloc[start_index, 2:5] = call_API(df,start_index)
        except Exception as e:
            print(f"Error: {e}")

    return df

### Un petit POC (pas nécessaire pour la suite)

### sauter cette partie si les fichiers csv existent déjà dans temporary_data/plantes

In [ ]:
remplir_tableau(plantes_sc_labelled, 7)

In [ ]:
plantes_sc_labelled.head(15) # verification des 7 premières lignes remplies et 3 suivantes non remplies

In [ ]:
# (uniquement utile pour faire des tests) vide les colonnes "Famille", "Genre" et "Espece" du dataframe plantes_sc_labelled
plantes_sc_labelled.iloc[:, 2:5] = ""

In [ ]:
plantes_sc_labelled.shape # verification du maintien de la dimension du dataframe : OK

## Pipeline

Il faut environ 3 secondes pour un call API, soit environ 2h10 pour les 2620 plantes à traiter dans plantes_sc_labelled.

On va donc stocker le dataframe dans un csv entre chaque appel à l'API pour ne pas perdre le travail effectué.

In [ ]:
def save_dataframe(df, df_name):
    df.to_csv(f"data/temporary_data/plantes/{df_name}.csv", index=False)

In [ ]:
def import_dataframe(df_name):
    return pd.read_csv(f"data/temporary_data/plantes/{df_name}.csv")

In [ ]:
plantes_sc_labelled = import_dataframe("plantes_sc_labelled")
plantes_precision_labelled = import_dataframe("plantes_precision_labelled")
plantes_precision_unlabelled = import_dataframe("plantes_precision_unlabelled")

In [ ]:
def pipeline_remplissage(df_name, num):
    df = import_dataframe(df_name)
    df[["Famille", "Genre", "Espece"]] = df[["Famille", "Genre", "Espece"]].fillna("")
    remplir_tableau(df, num)
    save_dataframe(df, df_name)

In [ ]:
# POC
# pipeline_remplissage("plantes_precision_unlabelled", 10)

### Sauvegarde progressive des calls API

In [ ]:
def pipeline(df_name, n):
    for i in range(n):
        try:
            pipeline_remplissage(df_name, 10) # une itération prend 30 secondes
        except Exception as e:
            print(f"Error occurred after {i} successful runs: {e}")
            break   

permet d'effectuer une sauvegarde des telechargements toutes les 10 occurences ie toutes les 30 secondes

In [ ]:
pipeline('plantes_precision_unlabelled', 1)

### Verification de la progression des téléchargements

In [ ]:
def verification_progression(df_name):
    # importe le dataframe plantes_sc_labelled depuis le fichier csv
    df = pd.read_csv(f"data/temporary_data/plantes/{df_name}.csv")
    # utilise missingno pour visualiser les valeurs manquantes dans le dataframe
    msno.matrix(df)
    
    # estimation de la quantité de travail restante en téléchargement de données
    num_nan = df["Famille"].isna().sum()
    percent_nan = round((num_nan / len(df)) * 100,2)
    print("\n")
    print(f"Nombre de lignes dans 'Famille' restants à télécharger : {num_nan}")
    print(f"Pourcentage de lignes dans 'Famille' restants à télécharger : {percent_nan} %")

In [ ]:
# importe plantes_precision_unlabelled depuis le fichier csv
plantes_precision_unlabelled = import_dataframe("plantes_precision_unlabelled")
# toutes les lignes de famille genre espece conteant "erreur_API" doivent être vidées avec ""
plantes_precision_unlabelled[["Famille", "Genre", "Espece"]] = plantes_precision_unlabelled[["Famille", "Genre", "Espece"]].replace("erreur_API", "")
# sauvegarde le dataprecisioname dans un fichier csv
save_dataframe(plantes_precision_unlabelled, "plantes_precision_unlabelled")


In [ ]:
verification_progression('plantes_precision_unlabelled')

## Export des résultats

In [ ]:
def export_plantes(nom_df):
    # Importer le dataframe
    df = pd.read_csv(f"data/temporary_data/plantes/{nom_df}.csv")

    # Vérifier si la colonne "Famille" contient des valeurs NaN
    if df["Famille"].isna().any():
        num_nan = df["Famille"].isna().sum()
        percent_nan = round((num_nan / len(df)) * 100,2)
        print("Il reste", percent_nan,"% des données à télécharger, soit", num_nan,"lignes.")
        msno.matrix(df)
    else:
        # Sauvegarder le dataframe dans un fichier CSV
        df.to_csv(f"data/plantes/{nom_df}.csv", index=False)

In [ ]:
export_plantes('plantes_precision_unlabelled')

## Décompression des données

On a à l'origine un dataset spipoll (670 744 lignes), groupé par collection_id (75 372 lignes) scindé en la partition plantes_sc_labelled et plantes_sc_unlabelled (respectivement 63 465 et 11 907 lignes). plantes_sc_unlabelled est scindé en la partition plantes_precision_labelled et plantes_precision_unlabelled (respectivement 9619 et 2288 lignes).


On avait compressé plantes_sc_labelled et plantes_precision_labelled en les groupant par valeur de plantes_sc et plantes_precision identiques (respectivement passé de 63 465 lignes à 2620 ; et passé de 9619 lignes à 4319). On va ici faire l'opération inverse pour compléter les lignes d'origine avec les valeurs obtenues par le traitement API.

In [479]:
# on definit les deux dataframes à reconstruire
plantes_sc_labelled_long = plantes_sc_labelled # issus des dataframes vierges du début du notebook avant compression ...
plantes_precision_labelled_long = plantes_precision_labelled # ... pas des dataframes complétés par l'API !!

print(plantes_sc_labelled_long.shape[0], plantes_precision_labelled_long.shape[0])

63465 9619


In [480]:
# ajout des colonnes "Famille", "Genre" et "Espece" dans les dataframes longs
plantes_sc_labelled_long.insert(2, "Famille", "")
plantes_sc_labelled_long.insert(3, "Genre", "")
plantes_sc_labelled_long.insert(4, "Espece", "")

plantes_precision_labelled_long.insert(2, "Famille", "")
plantes_precision_labelled_long.insert(3, "Genre", "")
plantes_precision_labelled_long.insert(4, "Espece", "")

In [481]:
# on trie les dataframes longs par plantes_sc et plantes_precision
plantes_sc_labelled_long = plantes_sc_labelled_long.sort_values(by="plante_sc")
plantes_precision_labelled_long = plantes_precision_labelled_long.sort_values(by="plante_precision")

In [482]:
plantes_sc_labelled_long.head()

collection_id        plante_sc Famille Genre Espece plante_fr  \
535659        59 851  Abelia triflora                            NaN   
670140        75 969  Abelia triflora                            NaN   
472291        52 148  Abelia triflora                            NaN   
472300        52 152  Abelia triflora                            NaN   
50579          5 567  Abelia triflora                            NaN   

       plante_precision  plante_inconnue plante_caractere  \
535659              NaN              0.0          plantée   
670140              NaN              0.0          plantée   
472291              NaN              NaN          plantée   
472300              NaN              NaN          plantée   
50579               NaN              NaN          plantée   

                                              photo_fleur  \
535659  https://spgp-api.65mo.fr/api/containers/spgp/d...   
670140  https://spgp-api.65mo.fr/api/containers/spgp/d...   
472291  https://spgp-api.65mo.fr/api/containers/spgp/d...   
472300  https://spgp-api.65mo.fr/api/containers/spgp/d...   
50579   https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                             photo_plante  \
535659  https://spgp-api.65mo.fr/api/containers/spgp/d...   
670140  https://spgp-api.65mo.fr/api/containers/spgp/d...   
472291  https://spgp-api.65mo.fr/api/containers/spgp/d...   
472300  https://spgp-api.65mo.fr/api/containers/spgp/d...   
50579   https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                            photo_feuille  
535659  https://spgp-api.65mo.fr/api/containers/spgp/d...  
670140  https://spgp-api.65mo.fr/api/containers/spgp/d...  
472291  https://spgp-api.65mo.fr/api/containers/spgp/d...  
472300  https://spgp-api.65mo.fr/api/containers/spgp/d...  
50579   https://spgp-api.65mo.fr/api/containers/spgp/d...

In [483]:
print(plantes_sc_labelled_long.shape[0], plantes_precision_labelled_long.shape[0])

63465 9619


In [484]:
# on importe les petits dataframes qui ont été complétés avec l'API précédemment
plantes_sc_labelled = pd.read_csv(f"data/plantes/plantes_sc_labelled.csv")
plantes_precision_labelled = pd.read_csv(f"data/plantes/plantes_precision_labelled.csv")

print(plantes_sc_labelled.shape[0], plantes_precision_labelled.shape[0])

2620 4314


### Petit POC pour construire la fonction d'intérêt

In [ ]:
# prend les 50 premières lignes du dataframe plantes_sc_labelled_long et mets les dans un df POC
plantes_sc_labelled_long_POC = plantes_sc_labelled_long.head(50)
plantes_sc_labelled_long_POC.shape

In [453]:
# Get the values of "Famille", "Genre", and "Espece" from the row in `plantes_sc_labelled` where the index is 856
famille, genre, espece = plantes_sc_labelled.loc[856, ["Famille", "Genre", "Espece"]]

# Identify the rows in `plantes_sc_labelled_long_POC` where "plante_sc" equals `plantes_sc_labelled["plante_sc"].iloc[856]`
mask = plantes_sc_labelled_long_POC["plante_sc"] == plantes_sc_labelled["plante_sc"].iloc[856]

# Replace the values of "Famille", "Genre", and "Espece" in these rows with the values from `plantes_sc_labelled`
plantes_sc_labelled_long_POC.loc[mask, "Famille"] = famille
plantes_sc_labelled_long_POC.loc[mask, "Genre"] = genre
plantes_sc_labelled_long_POC.loc[mask, "Espece"] = espece

In [ ]:
plantes_sc_labelled_long_POC.head(15)

### Construction de la fonction de décompression

In [485]:
def remplacer_valeurs(df_court, df_long, colonne_interet):
    # Boucle sur chaque ligne dans `df_court`
    for i in tqdm(range(len(df_court))):
        # Obtient les valeurs de "Famille", "Genre", et "Espece" de la ligne dans `df_court` où l'index est i
        famille, genre, espece = df_court.loc[i, ["Famille", "Genre", "Espece"]]

        # Identifie les lignes dans `df_long` où "plante_sc" est égal à `df_court["plante_sc"].iloc[i]`
        masque = df_long[f"{colonne_interet}"] == df_court[f"{colonne_interet}"].iloc[i]

        # Remplace les valeurs de "Famille", "Genre", et "Espece" dans ces lignes avec les valeurs de `df_court`
        df_long.loc[masque, "Famille"] = famille
        df_long.loc[masque, "Genre"] = genre
        df_long.loc[masque, "Espece"] = espece

In [ ]:
# Appelle la fonction avec `plantes_sc_labelled` comme `df_court` et `plantes_sc_labelled_long_POC` comme `df_long`
remplacer_valeurs(plantes_sc_labelled, plantes_sc_labelled_long_POC, "plante_sc")

In [ ]:
remplacer_valeurs(plantes_sc_labelled, plantes_sc_labelled_long, "plante_sc")

In [ ]:
remplacer_valeurs(plantes_precision_labelled, plantes_precision_labelled_long, "plante_precision")

In [524]:
# trie plantes_sc_labelled_long et plantes_precision_labelled_long par index croissant
plantes_sc_labelled_long = plantes_sc_labelled_long.sort_index()
plantes_precision_labelled_long = plantes_precision_labelled_long.sort_index()

In [533]:
plantes_sc_labelled_long.shape

(63465, 12)

In [534]:
plantes_precision_labelled_long.shape

(9619, 12)

In [527]:
plantes_precision_unlabelled = pd.read_csv(f"data/plantes/plantes_precision_unlabelled.csv")
plantes_precision_unlabelled.shape

(2288, 12)

In [531]:
# merge les dataframes plantes_sc_labelled_long, plantes_precision_labelled_long et plantes_precision_unlabelled
plantes_merged = pd.concat([plantes_sc_labelled_long, plantes_precision_labelled_long, plantes_precision_unlabelled])

# trie le dataframe plantes_merged par collection_id croissants
plantes_merged = plantes_merged.sort_values(by="collection_id")

# sauvegarde le dataframe plantes_merged dans le dossier temporary_data/plantes
plantes_merged.to_csv("data/temporary_data/plantes/plantes_merged.csv", index=False)

plantes_merged.shape

(75372, 12)

In [532]:
plantes_merged.head(100)

collection_id                                          plante_sc  \
0                1  Les Trèfles à fleurs blanches ou roses en boul...   
9               10             Les Ajoncs (des espèces du genre Ulex)   
65             100  Les Abricotiers, Pruniers et Amandiers (à fleu...   
66             101                      Les Pissenlits (Taraxacum sp)   
67             102  Les Véroniques à fleurs isolées le long de la ...   
...            ...                                                ...   
956         10 087                                    Plante inconnue   
8580        10 088                                 Borago officinalis   
957         10 090                                    Plante inconnue   
310         10 091                                    Plante inconnue   
8581        10 092                              Peucedanum ostruthium   

             Famille        Genre                  Espece  \
0           Fabaceae    Trifolium      Trifolium pratense   
9           Fabaceae  Adenocarpus  Adenocarpus foliolosus   
65          Rosaceae       Prunus          Prunus spinosa   
66        Asteraceae    Taraxacum      Taraxacum palustre   
67    Plantaginaceae     Veronica        Veronica persica   
...              ...          ...                     ...   
956         Rosaceae      Spiraea  Spiraea chamaedryfolia   
8580    Boraginaceae       Borago      Borago officinalis   
957       Asteraceae       Cynara      Cynara cardunculus   
310        Lamiaceae       Mentha       Mentha suaveolens   
8581        Apiaceae  Imperatoria  Imperatoria ostruthium   

                    plante_fr                       plante_precision  \
0                         NaN                                    NaN   
9                         NaN                 Ajonc (Ulex europaeus)   
65                        NaN                                    NaN   
66                        NaN       Pissenlit (Taraxacum officinale)   
67                        NaN  Véronique de Perse (Veronica persica)   
...                       ...                                    ...   
956                       NaN                             Pyracantha   
8580  La Bourrache officinale                                    NaN   
957                       NaN                              Artichaud   
310                       NaN                                    NaN   
8581     Impératoire, Benjoin                            impératoire   

      plante_inconnue    plante_caractere  \
0                 0.0           spontanée   
9                 0.0  ne se prononce pas   
65                0.0           spontanée   
66                0.0           spontanée   
67                0.0           spontanée   
...               ...                 ...   
956               0.0             plantée   
8580              0.0             plantée   
957               0.0             plantée   
310               0.0           spontanée   
8581              0.0           spontanée   

                                            photo_fleur  \
0     https://spgp-api.65mo.fr/api/containers/spgp/d...   
9     https://spgp-api.65mo.fr/api/containers/spgp/d...   
65    https://spgp-api.65mo.fr/api/containers/spgp/d...   
66    https://spgp-api.65mo.fr/api/containers/spgp/d...   
67    https://spgp-api.65mo.fr/api/containers/spgp/d...   
...                                                 ...   
956   https://spgp-api.65mo.fr/api/containers/spgp/d...   
8580  https://spgp-api.65mo.fr/api/containers/spgp/d...   
957   https://spgp-api.65mo.fr/api/containers/spgp/d...   
310   https://spgp-api.65mo.fr/api/containers/spgp/d...   
8581  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                           photo_plante  \
0     https://spgp-api.65mo.fr/api/containers/spgp/d...   
9     https://spgp-api.65mo.fr/api/containers/spgp/d...   
65    https://spgp-api.65mo.fr/api/containers/spgp/d...   
66    https://spgp-api.65mo.fr/api/container

## On va maintenant décompresser par collection_id

In [536]:
# Import spipoll.csv as a pandas dataframe
spipoll = pd.read_csv("././data/spipoll.csv",low_memory=False)

# Extract relevant columns from the spipoll dataframe
plantes = spipoll[['collection_id', 'plante_sc', 'plante_fr',
       'plante_precision', 'plante_inconnue', 'plante_caractere',
       'photo_fleur', 'photo_plante', 'photo_feuille']]

# ajout des colonnes "Famille", "Genre" et "Espece" dans le dataframe plantes
plantes.insert(2, "Famille", "")
plantes.insert(3, "Genre", "")
plantes.insert(4, "Espece", "")

In [537]:
plantes.shape

(670744, 12)

In [539]:
# import de plantes_merged depuis le fichier csv
plantes_merged = pd.read_csv("data/temporary_data/plantes/plantes_merged.csv")
plantes_merged.shape

(75372, 12)

In [541]:
remplacer_valeurs(plantes_merged, plantes, "collection_id")

100%|██████████| 75372/75372 [1:06:03<00:00, 19.02it/s]


In [542]:
plantes.head()

collection_id                                          plante_sc  \
0             1  Les Trèfles à fleurs blanches ou roses en boul...   
1             1  Les Trèfles à fleurs blanches ou roses en boul...   
2             2                                    Acanthus mollis   
3             2                                    Acanthus mollis   
4             2                                    Acanthus mollis   

       Famille      Genre              Espece      plante_fr plante_precision  \
0     Fabaceae  Trifolium  Trifolium pratense            NaN              NaN   
1     Fabaceae  Trifolium  Trifolium pratense            NaN              NaN   
2  Acanthaceae   Acanthus   Acanthus spinosus  Acanthe molle              NaN   
3  Acanthaceae   Acanthus   Acanthus spinosus  Acanthe molle              NaN   
4  Acanthaceae   Acanthus   Acanthus spinosus  Acanthe molle              NaN   

   plante_inconnue plante_caractere  \
0              0.0        spontanée   
1              0.0        spontanée   
2              NaN          plantée   
3              NaN          plantée   
4              NaN          plantée   

                                         photo_fleur  \
0  https://spgp-api.65mo.fr/api/containers/spgp/d...   
1  https://spgp-api.65mo.fr/api/containers/spgp/d...   
2  https://spgp-api.65mo.fr/api/containers/spgp/d...   
3  https://spgp-api.65mo.fr/api/containers/spgp/d...   
4  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                        photo_plante  \
0  https://spgp-api.65mo.fr/api/containers/spgp/d...   
1  https://spgp-api.65mo.fr/api/containers/spgp/d...   
2  https://spgp-api.65mo.fr/api/containers/spgp/d...   
3  https://spgp-api.65mo.fr/api/containers/spgp/d...   
4  https://spgp-api.65mo.fr/api/containers/spgp/d...   

                                       photo_feuille  
0  https://spgp-api.65mo.fr/api/containers/spgp/d...  
1  https://spgp-api.65mo.fr/api/containers/spgp/d...  
2  https://spgp-api.65mo.fr/api/containers/spgp/d...  
3  https://spgp-api.65mo.fr/api/containers/spgp/d...  
4  https://spgp-api.65mo.fr/api/containers/spgp/d...

In [543]:
# save plantes in a csv file in data/plantes
plantes.to_csv("data/plantes/plantes.csv", index=False)